### ISSUE TITLE
[Spree Upgrade][FrontOffice] Fix checkout feature specs (2 broken tests)

### ISSUE THREAD

**Issue by HugsDaniel** [openfoodfoundation/openfoodnetwork#2999]
## Description

Several specs are failing in `spec/features/consumer/shopping/checkout_spec.rb`, we have to fix them.
This is a first placeholder for these specs, I may open several issues later on.

## Expected Behavior

Specs should pass.

## Actual Behavior

Specs fail.

**Comment by luisramos0** [comment]
these are the 11 broken tests. I have investigated 3 of them but I have not found the fix yet:

```
  1) As a consumer I want to check out my cart with shipping and payment methods when I have an out of stock product in my cart returns me to the cart with an error message
     Failure/Error: page.should_not have_selector 'closing', text: "Checkout now"
       expected not to find visible css "closing" with text "Checkout now", found 1 match: "Checkout now"
     # ./spec/features/consumer/shopping/checkout_spec.rb:57:in `block (4 levels) in <top (required)>'
```

It's showing the buttons "login or checkout as guest"
Error "item unavailable" should be shown before login verification?

```
 2) As a consumer I want to check out my cart with shipping and payment methods login in as user with details filled out allows user to save default billing address and shipping address
     Failure/Error: order.bill_address.should be_nil
     
       expected: nil
            got: #<Spree::Address id: 24, firstname: "John", lastname: "Doe", address1: "10 Lovely Street", address2: ...-16 00:06:59", updated_at: "2018-11-16 00:06:59", company: "Company", latitude: nil, longitude: nil>
     # ./spec/features/consumer/shopping/checkout_spec.rb:107:in `block (5 levels) in <top (required)>'
```

order.bill_address should be nil before save
Why is bill address being saved?

```
 3) As a consumer I want to check out my cart with shipping and payment methods login in as user when the user has a preset shipping and billing address checks out successfully
     Failure/Error: @exception || "expected #{klass} to be enqueued matching #{options.inspect} (#{@jobs_created.count} others enqueued)"
     
     NoMethodError:
       undefined method `count' for nil:NilClass
     # ./spec/support/delayed_job_helper.rb:63:in `block (2 levels) in <module:DelayedJobHelper>'
     # ./spec/features/consumer/shopping/checkout_spec.rb:155:in `block (5 levels) in <top (required)>'
```

order_decorator. deliver_order_confirmation_email is calling DelayedJob but either job is not being enqueued or there’s a problem in the matcher in DelayedJobHelper.

```
4) As a consumer I want to check out my cart with shipping and payment methods login in as user with Stripe allows use of a saved card
     Failure/Error: expect(page).to have_content "Your order has been processed successfully"
       expected to find text "Your order has been processed successfully" in "SHOPS MAP PRODUCERS GROUPS ABOUT English Castellano English Account (arnaldo.shanahan@rice.info) Logout Shopping @ ENTERPRISE 41 0 items Enterprise 41 Herndon ABOUT ENTERPRISE 41 PRODUCERS CONTACT GROUPS Order #R151732123 Confirmed PAID Total order $11.23 Paying via: Stripe Billing address Will Marshall 123 Your Head, Melbourne, 3066, Victoria 0468363090 Frogs Delivery time time yellow Delivery address Will Marshall 123 Your Head, Melbourne, 3066, Victoria 0468363090 Item Price Qty Total Product #19 - 4923 (1g) $11.23 1 $11.23 Produce $11.23 Total $11.23 (includes tax) $1.02 Back To Store Interested in getting on the Open Food Network? Register here OFN Global Home | News | About | Contact OFN Sites Developer | Community | User Guide Secure and trusted. Open Food Network uses SSL encryption (2048 bit RSA) everywhere to keep your shopping and payment information private. Our servers do not store your credit card details and payments are processed by PCI-compliant services. Keep in touch Email us Newsletter sign-up News Calendar Navigate Shops Map Producers Groups About Join us Create a listing, shop or group directory on the Open Food Network. Tell me more! Read our Terms and conditions | Find us on GitHub Open Food Network is a free and open source software platform. Our content is licensed with CC BY-SA 3.0 and our code with AGPL 3. We take good care of your data. See our cookies policy"
     # ./spec/features/consumer/shopping/checkout_spec.rb:209:in `block (5 levels) in <top (required)>'

  5) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing takes us to the order confirmation page when we submit a complete form
     Failure/Error: @exception || "expected #{klass} to be enqueued matching #{options.inspect} (#{@jobs_created.count} others enqueued)"
     
     NoMethodError:
       undefined method `count' for nil:NilClass
     # ./spec/support/delayed_job_helper.rb:63:in `block (2 levels) in <module:DelayedJobHelper>'
     # ./spec/features/consumer/shopping/checkout_spec.rb:361:in `block (5 levels) in <top (required)>'

  6) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing with basic details filled takes us to the order confirmation page when submitted with 'same as billing address' checked
     Failure/Error: page.should have_content "Your order has been processed successfully"
       expected to find text "Your order has been processed successfully" in "SHOPS MAP PRODUCERS GROUPS ABOUT English Castellano English Login Shopping @ ENTERPRISE 80 0 items Enterprise 80 Herndon ABOUT ENTERPRISE 80 PRODUCERS CONTACT GROUPS Order #R257276274 Confirmed NOT PAID Total order $11.23 Paying via: Roger rabbit Billing address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Frogs Delivery time time yellow Delivery address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Item Price Qty Total Product #32 - 5034 (1g) $11.23 1 $11.23 Produce $11.23 Total $11.23 (includes tax) $1.02 Back To Store Interested in getting on the Open Food Network? Register here OFN Global Home | News | About | Contact OFN Sites Developer | Community | User Guide Secure and trusted. Open Food Network uses SSL encryption (2048 bit RSA) everywhere to keep your shopping and payment information private. Our servers do not store your credit card details and payments are processed by PCI-compliant services. Keep in touch Email us Newsletter sign-up News Calendar Navigate Shops Map Producers Groups About Join us Create a listing, shop or group directory on the Open Food Network. Tell me more! Read our Terms and conditions | Find us on GitHub Open Food Network is a free and open source software platform. Our content is licensed with CC BY-SA 3.0 and our code with AGPL 3. We take good care of your data. See our cookies policy"
     # ./spec/features/consumer/shopping/checkout_spec.rb:410:in `block (6 levels) in <top (required)>'

  7) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing with basic details filled takes us to the cart page with an error when a product becomes out of stock just before we purchase
     Failure/Error: page.should have_selector 'closing', text: "Your shopping cart"
       expected to find visible css "closing" with text "Your shopping cart" but there were no matches
     # ./spec/features/consumer/shopping/checkout_spec.rb:420:in `block (6 levels) in <top (required)>'

  8) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing with basic details filled when we are charged a shipping fee creates a payment for the full amount inclusive of shipping
     Failure/Error: page.should have_content "Your order has been processed successfully"
       expected to find text "Your order has been processed successfully" in "SHOPS MAP PRODUCERS GROUPS ABOUT English Castellano English Login Shopping @ ENTERPRISE 98 0 items Enterprise 98 Herndon ABOUT ENTERPRISE 98 PRODUCERS CONTACT GROUPS Order #R788678553 Confirmed NOT PAID Total order $11.23 Paying via: Roger rabbit Billing address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Frogs Delivery time time yellow Delivery address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Item Price Qty Total Product #38 - 1264 (1g) $11.23 1 $11.23 Produce $11.23 Total $11.23 (includes tax) $1.02 Back To Store Interested in getting on the Open Food Network? Register here OFN Global Home | News | About | Contact OFN Sites Developer | Community | User Guide Secure and trusted. Open Food Network uses SSL encryption (2048 bit RSA) everywhere to keep your shopping and payment information private. Our servers do not store your credit card details and payments are processed by PCI-compliant services. Keep in touch Email us Newsletter sign-up News Calendar Navigate Shops Map Producers Groups About Join us Create a listing, shop or group directory on the Open Food Network. Tell me more! Read our Terms and conditions | Find us on GitHub Open Food Network is a free and open source software platform. Our content is licensed with CC BY-SA 3.0 and our code with AGPL 3. We take good care of your data. See our cookies policy"
     # ./spec/features/consumer/shopping/checkout_spec.rb:429:in `block (7 levels) in <top (required)>'

  9) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing with basic details filled when we are charged a payment method fee (transaction fee) creates a payment including the transaction fee
     Failure/Error: expect(page).to have_content "Your order has been processed successfully"
       expected to find text "Your order has been processed successfully" in "SHOPS MAP PRODUCERS GROUPS ABOUT English Castellano English Login Shopping @ ENTERPRISE 107 0 items Enterprise 107 Herndon ABOUT ENTERPRISE 107 PRODUCERS CONTACT GROUPS Order #R085007611 Confirmed NOT PAID Total order $16.90 Paying via: Check Billing address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Frogs Delivery time time yellow Delivery address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Item Price Qty Total Product #41 - 3756 (1g) $11.23 1 $11.23 Produce $11.23 Transaction fee $5.67 Total $16.90 (includes tax) $1.02 Back To Store Interested in getting on the Open Food Network? Register here OFN Global Home | News | About | Contact OFN Sites Developer | Community | User Guide Secure and trusted. Open Food Network uses SSL encryption (2048 bit RSA) everywhere to keep your shopping and payment information private. Our servers do not store your credit card details and payments are processed by PCI-compliant services. Keep in touch Email us Newsletter sign-up News Calendar Navigate Shops Map Producers Groups About Join us Create a listing, shop or group directory on the Open Food Network. Tell me more! Read our Terms and conditions | Find us on GitHub Open Food Network is a free and open source software platform. Our content is licensed with CC BY-SA 3.0 and our code with AGPL 3. We take good care of your data. See our cookies policy"
     # ./spec/features/consumer/shopping/checkout_spec.rb:451:in `block (7 levels) in <top (required)>'

  10) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing with basic details filled credit card payments with a credit card payment method using Spree::Gateway::Bogus takes us to the order confirmation page when submitted with a valid credit card
      Failure/Error: page.should have_content "Your order has been processed successfully"
        expected to find text "Your order has been processed successfully" in "SHOPS MAP PRODUCERS GROUPS ABOUT English Castellano English Login Shopping @ ENTERPRISE 116 0 items Enterprise 116 Herndon ABOUT ENTERPRISE 116 PRODUCERS CONTACT GROUPS Order #R502584680 Confirmed PAID Total order $11.23 Paying via: Roger rabbit Billing address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Frogs Delivery time time yellow Delivery address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Item Price Qty Total Product #44 - 2870 (1g) $11.23 1 $11.23 Produce $11.23 Total $11.23 (includes tax) $1.02 Back To Store Interested in getting on the Open Food Network? Register here OFN Global Home | News | About | Contact OFN Sites Developer | Community | User Guide Secure and trusted. Open Food Network uses SSL encryption (2048 bit RSA) everywhere to keep your shopping and payment information private. Our servers do not store your credit card details and payments are processed by PCI-compliant services. Keep in touch Email us Newsletter sign-up News Calendar Navigate Shops Map Producers Groups About Join us Create a listing, shop or group directory on the Open Food Network. Tell me more! Read our Terms and conditions | Find us on GitHub Open Food Network is a free and open source software platform. Our content is licensed with CC BY-SA 3.0 and our code with AGPL 3. We take good care of your data. See our cookies policy"
      # ./spec/features/consumer/shopping/checkout_spec.rb:473:in `block (9 levels) in <top (required)>'

  11) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing with basic details filled credit card payments with a credit card payment method using Spree::Gateway::BogusSimple takes us to the order confirmation page when submitted with a valid credit card
      Failure/Error: page.should have_content "Your order has been processed successfully"
        expected to find text "Your order has been processed successfully" in "SHOPS MAP PRODUCERS GROUPS ABOUT English Castellano English Login Shopping @ ENTERPRISE 128 0 items Enterprise 128 Herndon ABOUT ENTERPRISE 128 PRODUCERS CONTACT GROUPS Order #R053644630 Confirmed PAID Total order $11.23 Paying via: Roger rabbit Billing address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Frogs Delivery time time yellow Delivery address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Item Price Qty Total Product #48 - 4006 (1g) $11.23 1 $11.23 Produce $11.23 Total $11.23 (includes tax) $1.02 Back To Store Interested in getting on the Open Food Network? Register here OFN Global Home | News | About | Contact OFN Sites Developer | Community | User Guide Secure and trusted. Open Food Network uses SSL encryption (2048 bit RSA) everywhere to keep your shopping and payment information private. Our servers do not store your credit card details and payments are processed by PCI-compliant services. Keep in touch Email us Newsletter sign-up News Calendar Navigate Shops Map Producers Groups About Join us Create a listing, shop or group directory on the Open Food Network. Tell me more! Read our Terms and conditions | Find us on GitHub Open Food Network is a free and open source software platform. Our content is licensed with CC BY-SA 3.0 and our code with AGPL 3. We take good care of your data. See our cookies policy"
      # ./spec/features/consumer/shopping/checkout_spec.rb:473:in `block (9 levels) in <top (required)>'
```

**Comment by luisramos0** [comment]
Feel free to pick any of the remaining broken tests. I think I'll look at something else now :-)

**Comment by HugsDaniel** [comment]
Two failures remain, which seem related to shipping tax calculations : 
```
1) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing takes us to the order confirmation page when we submit a complete form
     Failure/Error: expect(page).to have_content with_currency(1.93)
       expected to find text "$1.93" in "SHOPS MAP PRODUCERS GROUPS ABOUT English Castellano English Français Login Shopping @ ENTERPRISE 47 0 items Enterprise 47 Herndon ABOUT ENTERPRISE 47 PRODUCERS CONTACT GROUPS Order #R883562774 Confirmed NOT PAID Total order $11.23 Paying via: Roger rabbit Billing address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Frogs Delivery time time yellow Delivery address Will Marshall 123 Your Face, Melbourne, 3066, Victoria 0468363090 Your notes: SpEcIaL NoTeS Item Price Qty Total Product #21 - 6766 (1g) $11.23 1 $11.23 Produce $11.23 Total $11.23 (includes tax) $1.02 Back To Store Interested in getting on the Open Food Network? Register here OFN Global Home | News | About | Contact OFN Sites Developer | Community | User Guide Secure and trusted. Open Food Network uses SSL encryption (2048 bit RSA) everywhere to keep your shopping and payment information private. Our servers do not store your credit card details and payments are processed by PCI-compliant services. Keep in touch Email us Newsletter sign-up News Calendar Navigate Shops Map Producers Groups About Join us Create a listing, shop or group directory on the Open Food Network. Tell me more! Read our Terms and conditions | Find us on GitHub Open Food Network is a free and open source software platform. Our content is licensed with CC BY-SA 3.0 and our code with AGPL 3. We take good care of your data. See our cookies policy"
     # ./spec/features/consumer/shopping/checkout_spec.rb:380:in `block (5 levels) in <top (required)>'
     # -e:1:in `<main>'
```

```
 2) As a consumer I want to check out my cart with shipping and payment methods on the checkout page with payments open purchasing with basic details filled when we are charged a shipping fee creates a payment for the full amount inclusive of shipping
     Failure/Error: o.adjustments.shipping.first.amount.should == 4.56
     
       expected: 4.56
            got: 0.0 (#<BigDecimal:5604ec3aa470,'0.0',9(18)>) (using ==)
     # ./spec/features/consumer/shopping/checkout_spec.rb:435:in `block (7 levels) in <top (required)>'
     # -e:1:in `<main>'
```

I'm not really sure how to tackle these ones. @luisramos0 @sauloperez do you master these adjustment calculation processes ? I'd be up for a quick explanation so I can solve these two.

**Comment by luisramos0** [comment]
Hey @HugsDaniel I think it's better to leave these 2 with shipping fees for later, after #2683.
Actually, if we rebase #2775 after your #3100 is merged we may see these will be fixed already :-)

**Comment by HugsDaniel** [comment]
Hey @luisramos0 can you try rebasing #2775 to see if #3100 fixed these specs ?

**Comment by luisramos0** [comment]
done, thanks for the heads up!

**Comment by HugsDaniel** [comment]
![feelsgood](https://user-images.githubusercontent.com/28786869/49149159-a5664b80-f309-11e8-9d31-4b63e17f922c.png)

**Comment by sigmundpetersen** [comment]
So we can close? @luisramos0

**Comment by luisramos0** [comment]
no, only #2775 will close this. I need to write new specs as part of 2775 wip

**Comment by luisramos0** [comment]
closed by #2775
